### DISCUSSION TITLE
NextJS should provide a mechanism for altering asset urls

### DISCUSSION THREAD

**Discussion by msluther** [vercel/next.js#23964] [Ideas]
# Feature request

NextJS should provide a means to alter built asset urls before they're rendered into a document.

## Is your feature request related to a problem? Please describe.

Our primary use case here is we are building 2 versions of every css file `[file].css` and `[file].rtl.css` to provide smaller bundled asset to customers.  The former gets served for LTR (left-to-right) customers and the latter for RTL (right-to-left) customers.

We have also occasionally had a need to split css by locale (e.g. `en-US` gets a different css file from `de-DE`)

## Describe the solution you'd like

NextJS's `<Head>` element (and other elements) to support accepting a function to alter file urls before they're rendered into the page, or a mechanism to inject functionality to alter file urls before they end up in the `DocumentContext.Provider`.

Ideally we could do this for any static asset, though our primary use case is css files.

## Describe alternatives you've considered

We're currently planning to use a work-around that wraps the `<Head>` element and reaches deep into nextjs to be able to rework the context. We realize this isn't ideal and we might get broken by a future change, thus this issue. 

Our wrapping component looks roughly like this:

```jsx
import React from 'react';
import { Head } from 'next/document';
import { DocumentContext } from 'next/dist/next-server/lib/document-context';

export default function RewritingHead({ fileRewriter, ...props }) {
  if (fileRewriter) {
    return <DocumentContext.Consumer>
      { ctx => {
        const { files } = ctx._documentProps;
        const newCtx = {
          ...ctx,
          _documentProps: {
            ...ctx._documentProps,
            files: files && files.map(fileRewriter)
          }
        };
        return <DocumentContext.Provider value={ newCtx }>
          <Head { ...props } />
        </DocumentContext.Provider>;
      }}
    </DocumentContext.Consumer>;
  }

  return <Head { ...props } />;
}
```

## Additional context

We're using the [`webpack-rtl-plugin`](https://github.com/romainberger/webpack-rtl-plugin) to automatically output the LTR & RTL css files.

**Comment by xessam** [comment]
@msluther did you find a solution for this one? 
i have this problem with using RTL css assets too.

**Comment by msluther** [comment]
@xessam We've basically implemented the custom `RewritingHead` component as above in the alternatives in our app and then render that rather than the default next/document component.  It is definitely ugly though, but a workable solution in the interim. We're just supplying the `fileRewriter` in props to this component when the document should be in rtl mode and have the `*.rtl.css` files. And the `fileRewriter` for the rtl case is effectively:

```js
function rewriteRtlCss(filename) {
  // If it's not a css file don't touch it.
  if (!/.css$/.test(filename)) return filename;

  // If it's a *.css or a *.min.css file then rewrite to *.rtl.css or *.rtl.min.css
  return filename.replace(/(.min.css|.css)$/, '.rtl$1');
}
```

Ours ended up a bit more complicated than that for other reasons, but ends up effectively the same (note: I hand rewrote that and haven't run it to try to simplify, so you may need to tweak it if you use it).

**Comment by msluther** [comment]
It looks like due to #16184 our work around we had, no longer works.

Is there a supported way to swap out files based on individual requests (in this case to render different css files so that we're not including both RTL & LTR css in one giant file)?
